### PR TITLE
test: including unit testing for search deployments state transition logic and model conversions

### PR DESCRIPTION
**Pull request by AgustinBettati**
## Description

Jira ticket: [INTMDB-1309](https://jira.mongodb.org/browse/INTMDB-1309)

Adds unit testing for search_deployment resource. As part of this the following tests were defined:

- test cases for logic that waits for deployment state to transition to valid state after create/update/delete operations.
- test cases for converting from terraform model to SDK model and vice versa.

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.

### CODE REVIEW

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (AgustinBettati) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: Having a separate file for this function, and its corresponding _unit.go file would be ideal here. Having more granular files will be viable if we separate implementation into packages (https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1643)

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (AgustinBettati) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: Detected a bug here thanks to test case. Not very relevant, simply a logic that continues the retry mechanism if a "Service Unavailable" is returned from API.

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (marcosuma) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: nit - just to align with the syntax below - said that, why did we use different config for the delete? why not just aligning it to all?
```suggestion
func createUpdateRetryTimeConfig(configuredTimeout time.Duration) retrystrategy.TimeConfig {
```

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (marcosuma) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: yes! I wonder if we can move forward with that change

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_state_transition_test.go
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment_state_transition_test.go]: 👀 
```suggestion
func TestSearchDeploymentStateTransitionForDelete(t *testing.T) {
```

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_state_transition_test.go
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment_state_transition_test.go]: 👀 
```suggestion
func TestSearchDeploymentStateTransition(t *testing.T) {
```

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (marcosuma) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: ah, so nice to read this - music to our ears :)

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (marcosuma) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: if we move forward with (#1643) this can become a separate file.

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: let's keep using reference here
```suggestion
func searchDeploymentRefreshFunc(ctx context.Context, projectID, clusterName string, client *SearchDeploymentService) retry.StateRefreshFunc {
```

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (andreaangiolillo) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: What was the strategy before? the name `RetryStrategyPendingState` is not super clear 🤔

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: Was unit testing the only reason why you define the interface here?

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go]: should we print what was the error to help with the debug?

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go
Comment (marcosuma) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go]: nit: we don't have a team guideline hence so far it's up to engineer's preference. My suggestion is to avoid using separate method to "optimize" code in unit testing. Like Google does, we should maximize in-test context without deferring to separate methods.

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go]: Maybe you could define a list of const to reuse in your file

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go
Comment (marcosuma) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go]: you don't have to follow this, I just wanted to bring this up.

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go
Comment (marcosuma) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go]: worth it adding other test cases with not all fields?

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go
Comment (marcosuma) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go]: same here... more test cases?

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go]: 👀 same in adding more info on the reason why

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_state_transition_test.go
Comment (andreaangiolillo) [inline mongodbatlas/fw_resource_mongodbatlas_search_deployment_state_transition_test.go]: Can't you reuse `stateTransitionTestCase`? in other work, do you need two structs or only 1?

**Review by andreaangiolillo — COMMENTED** [review]
Left some comments that are not blocking the PR

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: `RetryStrategyPendingState` is used for encryption at rest resource, but is not a valid state for the search deployment. The correct state for representing that a change is in progress is with `RetryStrategyUpdatingState`.

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: revised this and defined into a single function. The only difference is with the `MinTimeout` value, had defined a lower value for the delete operation as I could see during testing this operation was significantly faster. However I do imagine this depends a lot on the nature of each deployment.

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go]: I see your point. I extracted this method mainly for readability of the test case:
```
expectedTFModel: &tfSearchDeploymentRSModel{
  ID:          types.StringValue("111111111111111111111111"),
  ClusterName: types.StringValue("Cluster0"),
  ProjectID:   types.StringValue("222222222222222222222222"),
  StateName:   types.StringValue("IDLE"),
  Specs:       tfSpecsList(t, "S20_HIGHCPU_NVME", 2),
}
```
Unfortunately there is no way to define a terraform `types.List` inlined.

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go]: In the case of search deployment there are no optional parameters that can be provided. Everything is required, as well as spec lists which has to be of 1 element (validated by Atlas API).

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_state_transition_test.go
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment_state_transition_test.go]: :) thanks

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: SearchDeploymentService is an interface type, which in turn is a pointer to struct that fulfills the interface (instantiated in `NewSearchDeploymentServiceFromClient`). More details of why pointer to interfaces is discouraged: https://stackoverflow.com/a/44372954

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: Yes. If we define a separate package structure can see this defined in a separate file to avoid noise in the main implementation file. Did you have any particular concern?

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go]: makes sense, adjusted

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment_state_transition_test.go
Comment (AgustinBettati) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment_state_transition_test.go]: Yep was able to reuse here. Simply do not use/define the field `expectedResult *admin.ApiSearchDeploymentResponse` for delete tests. Thanks!

File: mongodbatlas/fw_resource_mongodbatlas_search_deployment.go
Comment (andreaangiolillo) [reply mongodbatlas/fw_resource_mongodbatlas_search_deployment.go]: no, I was wondering if I was missing other reasons. I agree that a separate package structure would help here

**Review by marcosuma — APPROVED** [review]
LGTM - approve to unblock

**Review by lantoli — APPROVED** [review]
very nice!

**Pull request merged**
